### PR TITLE
Adds support for elixir sub-projects without the need for umbrella applications

### PR DESCRIPTION
**Pull request by gjsduarte**
This PR will change the **hex** file fetcher to include path dependencies from a monorepo main `mix.exs` file.
With this change, dependabot will now correctly update dependencies used in sub-projects, as long as they exist as path dependencies of your project.

Addresses https://github.com/dependabot/dependabot-core/issues/820

### CODE REVIEW

**Comment by gjsduarte** [comment]
https://github.com/dependabot/dependabot-core/issues/820

**Comment by gjsduarte** [comment]
Here is the content of the encoded `mix.exs`:

```
defmodule BankPlatform.Mixfile do
  use Mix.Project

  def project do
    [apps_path: "apps",
     build_embedded: Mix.env == :prod,
     start_permanent: Mix.env == :prod,
     version: "1.0.0-dev",
     source_url: "https://github.com/wojtekmach/acme_bank",
     name: "Acme Bank",
     docs: [source_ref: "HEAD", main: "main", assets: "docs", extras: ["docs/main.md"]],
     deps: deps(),
     aliases: aliases()]
  end

  defp deps do
    [{:ex_doc, github: "elixir-lang/ex_doc", branch: "master", only: :dev},
     {:bank, path: "apps/bank"}]
  end

  defp aliases do
    ["ecto.setup": ["ecto.create", "ecto.migrate", "ecto.seed"],
     "ecto.seed": ["run ./apps/bank/priv/repo/seeds.exs"],
     "ecto.reset": ["ecto.drop", "ecto.setup"],
     "test": ["ecto.create --quiet", "ecto.migrate", "test"]]
  end
end
```

**Comment by jurre** [comment]
Thanks @gjsduarte, this updates the FileFetcher to ensure we pull in the right files, but it's not clear to me if the rest of the update process actually works with this setup.

It'd be awesome if you could add a few tests for the `UpdateChecker` and `FileUpdater` classes to demonstrate this and to make sure that we don't regress there in the future. LMK if I can help!

So far looks great and thanks for tackling this 🙇

**Comment by gjsduarte** [comment]
Thanks for the feedback @jurre. I've added tests for the `UpdateChecker` and `FileUpdater` classes.

Basically, they ensure the rest of the process behaves the same as with umbrella applications.

Let me know what you think.

File: hex/lib/dependabot/hex/file_fetcher.rb [hunk 1]
```diff
@@ -44,16 +45,26 @@ def lockfile
...
         apps_path = mixfile.content.match(APPS_PATH_REGEX)&.
                     named_captures&.fetch("path")
         return [] unless apps_path
 
-        app_directories = repo_contents(dir: apps_path).
-                          select { |f| f.type == "dir" }.
-                          map { |f| File.join(apps_path, f.name) }
+        repo_contents(dir: apps_path).
+          select { |f| f.type == "dir" }.
+          map { |f| File.join(apps_path, f.name) } if apps_path
```
Comment (feelepxyz) [inline hex/lib/dependabot/hex/file_fetcher.rb:55]: Looks like there's already a guard against `apps_path` being nil on line 51
```suggestion
          map { |f| File.join(apps_path, f.name) }
```

File: hex/spec/dependabot/hex/update_checker_spec.rb [hunk 2]
```diff
@@ -532,6 +532,42 @@
...
+          name: "apps/dependabot_business/mix.exs",
+          content: fixture("mixfiles", "dependabot_business")
+        )
+      end
+      let(:sub_mixfile2) do
+        Dependabot::DependencyFile.new(
+          name: "apps/dependabot_web/mix.exs",
+          content: fixture("mixfiles", "dependabot_web")
+        )
+      end
```
Comment (feelepxyz) [inline hex/spec/dependabot/hex/update_checker_spec.rb:551]: Once you've got the project fixture set up this should work:
```suggestion
      let(:files) { project_dependency_files("umbrella_sub_projects") }
```

File: hex/spec/dependabot/hex/file_updater_spec.rb [hunk 3]
```diff
@@ -177,6 +177,81 @@
...
+            name: "apps/dependabot_business/mix.exs",
+            content: fixture("mixfiles", "dependabot_business")
+          )
+        end
+        let(:sub_mixfile2) do
+          Dependabot::DependencyFile.new(
+            name: "apps/dependabot_web/mix.exs",
+            content: fixture("mixfiles", "dependabot_web")
+          )
+        end
```
Comment (feelepxyz) [inline hex/spec/dependabot/hex/file_updater_spec.rb:197]: Same, once the project is set up: 
```suggestion
        let(:files) { project_dependency_files("umbrella_sub_projects") }
```

**Review by feelepxyz — COMMENTED** [review]
Thanks for the contribution 🙇  left a few minor comments otherwise LGTM!

**Comment by gjsduarte** [comment]
Thanks for the feedback @feelepxyz! I've refactored the tests to use the new pattern as you suggested.
Let me know what you think

**Review by feelepxyz — COMMENTED** [review]
Thanks for the spec changes 👌   I forgot to approve the action run so just waiting for ci to come back

**Review by feelepxyz — APPROVED** [review]
🙌

**Pull request merged**